content/docs/guides/parallel-execution.md

```diff
@@ -0,0 +1,380 @@
+---
+title: Parallel Execution
+subtitle: Using multiple threads to reduce test times
+polyglot:
+ - java
+ - kotlin
+ 
+weight: 1800
+---
+{{% block "java,kotlin" %}}
+Cucumber-JVM allows [parallel execution](https://cucumber.io/blog/announcing-cucumber-jvm-4-0-0/) across multiple threads since version 4.0.0. There are several options to incorporate this built-in feature in a Cucumber project.
+
+- [JUnit](/docs/guides/parallel-execution/#junit)
+- [TestNG](/docs/guides/parallel-execution/#testng)
+- [CLI](/docs/guides/parallel-execution/#cli)
+
+For each of these options, this tutorial will look at the project setup, configuration settings and execution commands.
+{{% /block %}}
+
+# JUnit
+
+{{% block "java,kotlin" %}}
+Cucumber can be executed in parallel using **JUnit and Maven test execution plugins**. In JUnit the **feature files are run in parallel rather than scenarios**, which means **all the scenarios in a feature file will be executed by the same thread**. You can use either Maven Surefire or Failsafe plugin to execute the runners.
+
+- Create a Maven project in your favorite IDE using the [cucumber-archetype](/docs/guides/10-minute-tutorial/#create-an-empty-cucumber-project) or adding Cucumber dependencies to the POM as detailed [here](https://cucumber.io/docs/installation/java/#maven) and Junit dependencies [here](/docs/cucumber/checking-assertions/#junit).
+
+- Create a **parallel** folder (or any other name) in `src/test/resources` path and add the two feature files (`scenarios.feature` and `scenario-outlines.feature`) inside it.
+
+```gherkin
+Feature: Scenarios feature file
+
+  Scenario: Scenario Number One
+    Given Step from 'Scenario 1' in 'scenarios' feature file
+
+  Scenario: Scenario Number Two
+    Given Step from 'Scenario 2' in 'scenarios' feature file
+```
+
+```gherkin
+Feature: Scenario Outlines feature file
+
+  Scenario Outline: <scen_out_row_num>
+    Given Step from '<scen_out_row_num>' in 'scenario-outlines' feature file
+
+    Examples: 
+      | scen_out_row_num       |
+      | Scenario Outline Row 1 |
+      | Scenario Outline Row 2 |
+```
+
+- Add the **step definition class** to the `parallel` package (same name as folder above for automatic pickup by runner) in `src/test/java` folder.
+{{% /block %}}
+
+{{% block "java" %}}
+
+```java
+package parallel;
+
+import cucumber.api.java.BeforeStep;
+import cucumber.api.java.en.Given;
+
+public class Stepdefs {
+
+	@Given("Step from {string} in {string} feature file")
+	public void step(String scenario, String file) {
+		System.out.format("Thread ID - %2d - %s from %s feature file.\n", Thread.currentThread().getId(), scenario,file);
+	}
+}
+```
+{{% /block %}}
+
+{{% block "kotlin" %}}
+
+```kotlin
+package parallel
+
+import cucumber.api.java8.En
+
+class Stepdefs : En {
+		
+		Given("Step from {string} in {string} feature file") { scenario: String , file: String ->

```

Comment:
```suggestion
		    init {
        Given("Step from {string} in {string} feature file") { scenario: String , file: String ->
            println("Thread ID - ${Thread.currentThread().id} - $scenario from $file feature file")
        }
    }
```

Reply:
When using lambda's with Kotlin, the stepdefs need to be inside an `init {}` block.
We could consider adding examples both with annotations and lambdas, but that could be a separate additional PR; let's get this one merged asap. :)

Reply:
Added to batch commit